clang/lib/Parse/ParseStmt.cpp

```diff
@@ -2427,3 +2447,47 @@ bool Parser::ParseOpenCLUnrollHintAttribute(ParsedAttributes &Attrs) {
   }
   return true;
 }
+
+bool Parser::ParseWhereClauseFact(WhereClause *WClause) {
+  SourceLocation Loc = Tok.getLocation();
+
+  if (Tok.is(tok::identifier) && NextToken().is(tok::colon)) {
+    IdentifierInfo *VarName = Tok.getIdentifierInfo();
+
+    // Consume the identifier.
+    ConsumeToken();
+    // Consume the ':' token.
+    ConsumeToken();
+
+    // Parse bounds expression.
+    ExprResult BoundsRes(ParseBoundsExpression());
+    return Actions.ActOnBoundsFact(WClause, VarName, BoundsRes,
+                                   getCurScope(), Loc);
+  }
+
+  ExprResult ExprRes(ParseExpression());
+  return Actions.ActOnRelopFact(WClause, ExprRes, Loc);
```

Comment:
Instead of passing WClause to functions ActOnBoundsFact and ActOnRelopFact, the functions could return a BoundsFact or a RelopFact (also, return null on error), which then gets added to WClause using the AddFact method of WClause. This will allow us to use ActOnBoundsFact and ActOnRelopFact in slightly different scenarios also.